src/analysis/bits-bounds-lattice.h

```diff
@@ -0,0 +1,144 @@
+#ifndef wasm_analysis_bits_bounds_lattice_h
+#define wasm_analysis_bits_bounds_lattice_h
+
+#include <optional>
+
+#include "lattice.h"
+#include "wasm.h"
+
+namespace wasm::analysis {
+
+struct MaxBitsLattice {
+  enum LatticeState { BOTTOM, REGULAR_VALUE, TOP };
+
+  struct Element {
+    Index upperBound = 0;
+    std::optional<Literal> constVal;
+    LatticeState state = BOTTOM;
+
+    bool isTop() const { return state == TOP; }
+
+    bool isBottom() const { return state == BOTTOM; }
+
+    void setUpperBound(Index upperBound) {
+      if (state != TOP) {
+        state = REGULAR_VALUE;
+        this->upperBound = upperBound;
+        constVal.reset();
+      }
+    }
+
+    void setUpperBound(Index upperBound, Literal constVal) {
+      if (state != TOP) {
+        state = REGULAR_VALUE;
+        this->upperBound = upperBound;
+        this->constVal = constVal;
```

Comment:
I'm surprised that both `upperBound` and `constVal` can be meaningful at once. Do you have an invariant that if both are present, then they must match? For example, is it the case that if `constVal` is `Literal(5)`, then `upperBound` must be 2 because the max bit of 5 is 2? If so, it might be simpler to only store the literal in that case and recalculate its max bits on the fly when necessary.